File: DKPlugins/DKDuktape/DKConsole.cpp
#ifdef USE_DKDuktape 
#include "DK/DKApp.h"
#include "DKDuktape/DKConsole.h"


bool DKConsole::Init(){
	DKDEBUGFUNC();
	DKDuktape::AttachFunction("CPP_DKConsole_assert", DKConsole::_assert);
	DKDuktape::AttachFunction("CPP_DKConsole_clear", DKConsole::clear);
	DKDuktape::AttachFunction("CPP_DKConsole_debug", DKConsole::debug);
	DKDuktape::AttachFunction("CPP_DKConsole_error", DKConsole::error);
	DKDuktape::AttachFunction("CPP_DKConsole_exception", DKConsole::exception);
	DKDuktape::AttachFunction("CPP_DKConsole_group", DKConsole::group);
	DKDuktape::AttachFunction("CPP_DKConsole_groupCollapsed", DKConsole::groupCollapsed);
	DKDuktape::AttachFunction("CPP_DKConsole_groupEnd", DKConsole::groupEnd);
	DKDuktape::AttachFunction("CPP_DKConsole_info", DKConsole::info);
	DKDuktape::AttachFunction("CPP_DKConsole_log", DKConsole::log);
	DKDuktape::AttachFunction("CPP_DKConsole_trace", DKConsole::trace);
	DKDuktape::AttachFunction("CPP_DKConsole_warn", DKConsole::warn);
	
	DKClass::DKCreate("DKDuktape/DKConsole.js");
	return true;
}

int DKConsole::_assert(duk_context* ctx){
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKERROR(string);
	return 1;
}

int DKConsole::clear(duk_context* ctx)
{
	//FIXME - make this work on all OS's
	DKString out;
	if(!DKUtil::System("cls", out)){ return 0; }
	return 1;
}

int DKConsole::debug(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKDEBUG(string);
	return 1;
}

int DKConsole::error(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKERROR(string);
	return 1;
}

int DKConsole::exception(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKERROR(string);
	return 1;
}

int DKConsole::group(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKINFO(string);
	return 1;
}

int DKConsole::groupCollapsed(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKINFO(string);
	return 1;
}

int DKConsole::groupEnd(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKINFO(string);
	return 1;
}

int DKConsole::info(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKINFO(string);
	return 1;
}

int DKConsole::log(duk_context* ctx)
{
	DKString string;
	if (duk_is_string(ctx, 0)) {
		string = duk_require_string(ctx, 0);
	}
	if (duk_is_boolean(ctx, 0)) {
		string = toString(duk_require_boolean(ctx, 0));
	}
	if (duk_is_number(ctx, 0)) {
		string = toString(duk_require_int(ctx, 0));
	}
	DKINFO(string);
	return 1;
}

int DKConsole::trace(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKWARN(string);
	return 1;
}

int DKConsole::warn(duk_context* ctx)
{
	DKString string;
	if(duk_is_string(ctx, 0)){
		string = duk_require_string(ctx, 0);
	}
	if(duk_is_boolean(ctx, 0)){
		string = toString(duk_require_boolean(ctx, 0));
	}
	if(duk_is_number(ctx, 0)){
		string = toString(duk_require_int(ctx, 0));
	}
	DKWARN(string);
	return 1;
}


#endif //USE_DKDuktape

File: DKPlugins/DKDuktape/DKConsole.js
// https://console.spec.whatwg.org/
// https://developer.mozilla.org/en-US/docs/Web/API/Console

const Console_HandleColor = function Console_HandleColor(msg){
	//TODO: handle coloring, (currently disabled)
	if(msg){
		msg = msg.replace("%c ","")
		msg = msg.replace("%c","")
	}
	return msg;
}

// https://console.spec.whatwg.org/#console-namespace
var Console = function Console(pointer) {
    Console.prototype.assert = function assert(msg) {
		msg = Console_HandleColor(msg)
        //if (assertion)
        //    return;
        CPP_DKConsole_assert(msg + "\n");
    }
    Console.prototype.clear = function clear() {
        CPP_DKConsole_clear();
    }
    Console.prototype.debug = function debug(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_debug(msg + "\n");
    }
    Console.prototype.error = function error(msg) {
		msg = Console_HandleColor(msg)
		if(!msg){
			CPP_DKConsole_error("msg invalid \n");
			return;
		}
		const err = new Error(msg);
		if (!err.stack) {
			try {
				throw err; // old browsers need the Error thrown to fill the stack
			} 
			catch (err) {
				if (!err.stack)
					return error("err.stack invalid"); // browser too old
			}
		}
		CPP_DKConsole_error(err.stack+"\n");
    }
    Console.prototype.exception = function exception(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_exception(msg + "\n");
    }
    Console.prototype.group = function group(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_group(msg + "\n");
    }
    Console.prototype.groupCollapsed = function groupCollapsed(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_groupCollapsed(msg + "\n");
    }
    Console.prototype.groupEnd = function groupEnd(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_groupEnd(msg + "\n");
    }
    Console.prototype.info = function info(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_info(msg + "\n");
    }
    Console.prototype.log = function log(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_log(msg + "\n");
    }
    Console.prototype.trace = function trace(msg) {
		msg = Console_HandleColor(msg)
		CPP_DKConsole_trace(msg + "\n");
    }
    Console.prototype.warn = function warn(msg) {
		msg = Console_HandleColor(msg)
        CPP_DKConsole_warn(msg + "\n");
    }
}



var console = new Console("console");
window.console = console;